File: app/src/main/java/br/com/mov/views/fragment/HomeFragment.java
package br.com.mov.views.fragment;

import android.graphics.Typeface;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.tabs.TabLayout;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Timer;
import java.util.TimerTask;

import br.com.mov.R;
import br.com.mov.models.Movie;
import br.com.mov.models.Slide;
import br.com.mov.views.recyclerview.adapter.MovieAdapter;
import br.com.mov.views.slidepager.adapter.SlidePagerAdapter;

public class HomeFragment extends Fragment {

    private List<Slide> slideList;
    private ViewPager sliderPager;
    private TabLayout indicator;
    private RecyclerView moviesRv;
    private RecyclerView moviesfiltrated;
    private Typeface font;
    private TextView firstMovies, secondMovies;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_home, container, false);
        getActivity().setTitle("Sotck Movie");

        font = Typeface.createFromAsset(getActivity().getAssets(), "fonts/OpenSans-Semibold.ttf");

        sliderPager = view.findViewById(R.id.view_pager);
        indicator = view.findViewById(R.id.indicator);
        moviesRv = view.findViewById(R.id.Rv_movies);
        moviesfiltrated = view.findViewById(R.id.Rv_movies_filtrates);
        firstMovies = view.findViewById(R.id.first_filtration);
        secondMovies = view.findViewById(R.id.second_filtration);
        firstMovies.setTypeface(font);
        secondMovies.setTypeface(font);

        slideList = new ArrayList<>();
        slideList.add(new Slide(R.drawable.slide1));
        slideList.add(new Slide(R.drawable.slide2));
        slideList.add(new Slide(R.drawable.slide1));
        slideList.add(new Slide(R.drawable.slide2));

        SlidePagerAdapter adapter = new SlidePagerAdapter(getContext(), slideList);
        sliderPager.setAdapter(adapter);

        Timer timer = new Timer();
        timer.scheduleAtFixedRate(new SliderTimer(), 4000, 6000);


        indicator.setupWithViewPager(sliderPager, true);

        List<Movie> movies = new ArrayList<>();
        movies.add(new Movie("Moana", R.drawable.moana));
        movies.add(new Movie("Black P", R.drawable.blackp));
        movies.add(new Movie("Perdido em Marte", R.drawable.themartian));
        movies.add(new Movie("Moana", R.drawable.moana));
        movies.add(new Movie("Black P", R.drawable.blackp));
        movies.add(new Movie("Perdido em Marte", R.drawable.themartian));
        movies.add(new Movie("Moana", R.drawable.moana));
        movies.add(new Movie("Black P", R.drawable.blackp));
        movies.add(new Movie("Perdido em Marte", R.drawable.themartian));
        movies.add(new Movie("Moana", R.drawable.moana));
        movies.add(new Movie("Black P", R.drawable.blackp));
        movies.add(new Movie("Perdido em Marte", R.drawable.themartian));

        List<Movie> moviesFitrates = new ArrayList<>();
        moviesFitrates.add(new Movie("Black P", R.drawable.blackp));
        moviesFitrates.add(new Movie("Black P", R.drawable.blackp));
        moviesFitrates.add(new Movie("Black P", R.drawable.blackp));
        moviesFitrates.add(new Movie("Black P", R.drawable.blackp));
        moviesFitrates.add(new Movie("Moana", R.drawable.moana));
        moviesFitrates.add(new Movie("Moana", R.drawable.moana));
        moviesFitrates.add(new Movie("Moana", R.drawable.moana));
        moviesFitrates.add(new Movie("Moana", R.drawable.moana));
        moviesFitrates.add(new Movie("Perdido em Marte", R.drawable.themartian));
        moviesFitrates.add(new Movie("Perdido em Marte", R.drawable.themartian));
        moviesFitrates.add(new Movie("Perdido em Marte", R.drawable.themartian));
        moviesFitrates.add(new Movie("Perdido em Marte", R.drawable.themartian));

        MovieAdapter movieAdapter = new MovieAdapter(getContext(), movies);
        MovieAdapter movieAdapter1 = new MovieAdapter(getContext(), moviesFitrates);
        moviesfiltrated.setAdapter(movieAdapter1);
        moviesRv.setAdapter(movieAdapter);
        moviesfiltrated.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
        moviesRv.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));

        return view;
    }

    class SliderTimer extends TimerTask {

        @Override
        public void run() {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT && getActivity() != null) {
                Objects.requireNonNull(getActivity()).runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        if (sliderPager.getCurrentItem() < slideList.size() - 1) {
                            sliderPager.setCurrentItem(sliderPager.getCurrentItem() + 1);
                        } else {
                            sliderPager.setCurrentItem(0);
                        }
                    }
                });
            }
        }
    }
}

File: app/src/main/java/br/com/mov/views/fragment/HomeFragment.kt
package br.com.mov.views.fragment

import android.graphics.Typeface
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.navigation.fragment.NavHostFragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import br.com.mov.R
import br.com.mov.models.Slide
import br.com.mov.views.recyclerview.adapter.MovieAdapter
import br.com.mov.views.slidepager.adapter.SlidePagerAdapter
import br.com.mov.views.viewmodel.LoginViewModel
import br.com.mov.views.viewmodel.MovieViewModel
import br.com.mov.views.viewmodel.StateAppComponentsViewModel
import br.com.mov.views.viewmodel.VisualComponents
import kotlinx.android.synthetic.main.fragment_home.*
import org.koin.android.viewmodel.ext.android.sharedViewModel
import org.koin.android.viewmodel.ext.android.viewModel
import java.util.*

class HomeFragment : Fragment() {

    private lateinit var slideList: ArrayList<Slide>
    private val appComponentsViewModel: StateAppComponentsViewModel by sharedViewModel()
    private val loginViewModel: LoginViewModel by viewModel()
    private val movieViewModel: MovieViewModel by viewModel()
    private lateinit var movieAdapter: MovieAdapter
    private lateinit var orderTitleMovieAdapter: MovieAdapter
    private lateinit var orderRatingMovieAdapter: MovieAdapter
    private val navController by lazy { findNavController(this) }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_home, container, false)

        checkStateLogin()
        appComponentsViewModel.havCoponent = VisualComponents(true)
        movieViewModel.getMovies()
        movieViewModel.getMoviesOrderByTitle()
        movieViewModel.getMoviesOrderByRating()
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val font = Typeface.createFromAsset(activity!!.assets, "fonts/OpenSans-Semibold.ttf")

        first_msg_movies.typeface = font
        msg_movies_order_title.typeface = font


        this.slideList = arrayListOf(
                Slide(R.drawable.batman_slide),
                Slide(R.drawable.interestellar_slide),
                Slide(R.drawable.starwars_slide),
                Slide(R.drawable.thorin_slide))

        val adapter = context?.let { SlidePagerAdapter(it, slideList) }
        slider_pager.adapter = adapter

        val timer = Timer()
        timer.scheduleAtFixedRate(SliderTimer(), 4000, 6000)

        indicator.setupWithViewPager(slider_pager, true)

        initMovieAdapter()
        initOrderTitleMovieAdapter()
        initOrderRatingMovieAdapter()
    }

    private fun initMovieAdapter() {
        movieViewModel.checkMoviesReturned()?.observe(viewLifecycleOwner,
                androidx.lifecycle.Observer { movieList ->
                    if (movieList != null) {
                        this.movieAdapter = context?.let { context ->
                            MovieAdapter(context, movieList)
                        }!!
                        rv_movies.adapter = movieAdapter
                        rv_movies.layoutManager = LinearLayoutManager(context,
                                LinearLayoutManager.HORIZONTAL, false)
                        movieAdapter.onItemCliclListener = { movie ->
                            goToMovieDetailFragment(movie.id)
                        }
                    }
                })
    }

    private fun initOrderTitleMovieAdapter() {
        movieViewModel.checkMoviesOrderByTitleReturned()?.observe(viewLifecycleOwner,
                androidx.lifecycle.Observer { movieList ->
                    if (movieList != null) {
                        this.orderTitleMovieAdapter = context?.let { context ->
                            MovieAdapter(context, movieList)
                        }!!
                        rv_movies_order_title.adapter = orderTitleMovieAdapter
                        rv_movies_order_title.layoutManager = LinearLayoutManager(context,
                                LinearLayoutManager.HORIZONTAL, false)
                        orderTitleMovieAdapter.onItemCliclListener = { movie ->
                            goToMovieDetailFragment(movie.id)
                        }
                    }
                })
    }

    private fun initOrderRatingMovieAdapter() {
        movieViewModel.checkMoviesOrderByRatingReturned()?.observe(viewLifecycleOwner,
                androidx.lifecycle.Observer { movieList ->
                    if (movieList != null) {
                        this.orderRatingMovieAdapter = context?.let { context ->
                            MovieAdapter(context, movieList)
                        }!!
                        rv_movies_order_rating.adapter = orderRatingMovieAdapter
                        rv_movies_order_rating.layoutManager = LinearLayoutManager(context,
                                LinearLayoutManager.HORIZONTAL, false)
                        orderRatingMovieAdapter.onItemCliclListener = { movie ->
                            goToMovieDetailFragment(movie.id)
                        }
                    }
                })
    }

    private fun checkStateLogin() {
        if (!loginViewModel.isLogged()) {
            goToLoginFragment()
        }
    }

    private fun goToLoginFragment() {
        val direction = HomeFragmentDirections
                .actionHomeFragmentToLoginFragment()
        navController.navigate(direction)
    }

    private fun goToMovieDetailFragment(movieId: Long?) {
        val id: Long = movieId!!
        val direction = HomeFragmentDirections.
                actionHomeFragmentToMovieDetailFragment(id)
        navController.navigate(direction)
    }

    internal inner class SliderTimer : TimerTask() {

        override fun run() {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT && activity != null && slider_pager != null) {
                Objects.requireNonNull<FragmentActivity>(activity).runOnUiThread {
                    if (slider_pager.currentItem < slideList.size - 1 && slider_pager != null) {
                        slider_pager.currentItem = slider_pager.currentItem + 1
                    } else {
                        slider_pager.currentItem = 0
                    }
                }
            }
        }
    }
}
